Source/Utils/Average.cpp

```diff
@@ -27,3 +27,99 @@ Average::ToCellCenter ( MultiFab& mf_out,
                      } );
     }
 }
+
+void
+Average::CoarsenAndInterpolateLoop ( MultiFab& mf_cp,
+                                     const MultiFab& mf_fp,
+                                     const int dcomp,
+                                     const int scomp,
+                                     const int ncomp,
+                                     const IntVect ratio )
+{
+    // Staggerings of input fine MultiFab and output coarse MultiFab
+    const IntVect stag_fp = mf_fp.boxArray().ixType().ixType();
+    const IntVect stag_cp = mf_cp.boxArray().ixType().ixType();
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.nGrowVect() >= IntVect(stag_cp-stag_fp),
+        "input MultiFab does not have enough guard cells for this interpolation" );
+
+    // Auxiliary integer arrays (always 3D unlike IntVect objects)
+    int sf[3], sc[3], cr[3];
+
+    sf[0] = stag_fp[0];
+    sf[1] = stag_fp[1];
+#if   (AMREX_SPACEDIM == 2)
+    sf[2] = 0;
+#elif (AMREX_SPACEDIM == 3)
+    sf[2] = stag_fp[2];
+#endif
+
+    sc[0] = stag_cp[0];
+    sc[1] = stag_cp[1];
+#if   (AMREX_SPACEDIM == 2)
+    sc[2] = 0;
+#elif (AMREX_SPACEDIM == 3)
+    sc[2] = stag_cp[2];
+#endif
+
+    cr[0] = ratio[0];
+    cr[1] = ratio[1];
+#if   (AMREX_SPACEDIM == 2)
+    cr[2] = 1;
+#elif (AMREX_SPACEDIM == 3)
+    cr[2] = ratio[2];
+#endif
```

Comment:
What about this? Trying to shorten and clarify.
```suggestion
    // Staggerings of input fine MultiFab and output coarse MultiFab
    IntVect sf = mf_fp.boxArray().ixType().ixType();
    IntVect sc = mf_cp.boxArray().ixType().ixType();
    IntVect cr = crse_ratio;
#if (AMREX_SPACEDIM == 2)
    sf[2] = 0;
    sc[2] = 0;
    cr[2] = 1;
#endif

    AMREX_ALWAYS_ASSERT_WITH_MESSAGE( mf_fp.nGrowVect() >= IntVect(stag_cp-stag_fp),
        "input MultiFab does not have enough guard cells for this interpolation" );
```

Reply:
I'm afraid that this cannot be done. The reason why I used C++ arrays instead of objects of type `IntVect` is that `IntVect` is a dimension-dependent class representing an integer vector in `AMREX_SPACEDIM`-dimensional space. Hence, when `AMREX_SPACEDIM=2` assignments like `sf[2]=0;` would try to access invalid memory locations (and be aborted by the ASSERT call in `amrex/Src/Base/AMReX_IntVect.H:186`). The reason for having this logic before calling the kernel that performs the interpolation was to write only one kernel algorithm in 3D that does not distinguish explicitly between 2D and 3D. Does it make sense? I will try to think about other ways to shorten and clarify this piece of code.